cmd/sharded-test-server/virtual.go

```diff
@@ -19,70 +19,239 @@ package main
-		fmt.Sprintf("--secure-port=%d", 7444+index),
 		"--requestheader-username-headers=X-Remote-User",
 		"--requestheader-group-headers=X-Remote-Group",
-		fmt.Sprintf("--requestheader-client-ca-file=%s", filepath.Join(workDirPath, ".kcp/requestheader-ca.crt")),
+		fmt.Sprintf("--requestheader-client-ca-file=%s", filepath.Join(v.workDirPath, ".kcp/requestheader-ca.crt")),
 		"--v=4",
 	)
 	fmt.Fprintf(out, "running: %v\n", strings.Join(commandLine, " "))
 
 	cmd := exec.CommandContext(ctx, commandLine[0], commandLine[1:]...) //nolint:gosec
 
-	logFilePath := filepath.Join(logDirPath, fmt.Sprintf("kcp-virtual-workspaces-%d.log", index))
+	logFilePath := filepath.Join(v.workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/virtualworkspace.log", v.index))
+	if v.logDirPath != "" {
+		logFilePath = filepath.Join(v.logDirPath, fmt.Sprintf("kcp-virtual-workspaces-%d.log", v.index))
+	}
+
 	if err := os.MkdirAll(filepath.Dir(logFilePath), 0755); err != nil {
-		return nil, err
+		return err
 	}
 	logFile, err := os.OpenFile(logFilePath, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	writer := helpers.NewHeadWriter(logFile, out)
-	cmd.Stdout = writer
+	v.writer = helpers.NewHeadWriter(logFile, out)
+	cmd.Stdout = v.writer
 	cmd.Stdin = os.Stdin
-	cmd.Stderr = writer
+	cmd.Stderr = v.writer
 
 	if err := cmd.Start(); err != nil {
-		return nil, err
+		return err
 	}
 
 	terminatedCh := make(chan error, 1)
+	v.terminatedCh = terminatedCh
 	go func() {
 		terminatedCh <- cmd.Wait()
 	}()
 
-	return terminatedCh, nil
+	return nil
+}
+
+func (v *VirtualWorkspace) waitForReady(ctx context.Context) (<-chan error, error) {
+	// wait for readiness
+	logger := klog.FromContext(ctx)
+	logger.WithValues("virtual-workspaces", v.index).Info("Waiting for virtual-workspaces /readyz to succeed")
+	for {
+		time.Sleep(100 * time.Millisecond)
+
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("context canceled")
+		case rc := <-v.terminatedCh:
+			return nil, fmt.Errorf("virtual-workspaces terminated with exit code %w", rc)
+		default:
+		}
+
+		vwHost := fmt.Sprintf("https://localhost:%d", 7444+v.index)
+		kubeconfigPath := filepath.Join(v.workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/virtualworkspace.kubeconfig", v.index))
+		vwConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfigPath},
+			// We override the Server here because virtualworkspace.kubeconfig is
+			// for VW->shard but we want to poll the VW endpoint readyz
+			&clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: vwHost}}).ClientConfig()
+		if err != nil {
+			logger.Error(err, "failed to create vw config")
+			continue
+		}
+
+		vwClient, err := kcpclientset.NewForConfig(vwConfig)
+		if err != nil {
+			logger.Error(err, "failed to create vw client")
+			continue
+		}
+
+		res := vwClient.RESTClient().Get().AbsPath("/readyz").Do(ctx)
```

Comment:
What does the output of this call look like? We have readiness checks for API-server-like binaries already, would be good to  reuse if this has e.g. post-start-hooks

Reply:
AFAICS this is consistent with the readyz checks elsewhere in the *test-server code, so perhaps we could consider going with this for now, then reworking all later so we don't end up with an inconsistent mix of approaches?

```
$ grep -R "RESTClient" ./*
./sharded-test-server/virtual.go:		res := vwClient.RESTClient().Get().AbsPath("/readyz").Do(ctx)
./sharded-test-server/frontproxy.go:		res := kcpClient.RESTClient().Get().AbsPath("/readyz").Do(ctx)
./sharded-test-server/cache.go:		res := cacheClient.RESTClient().Get().AbsPath("/readyz").Do(ctx)
./test-server/kcp/shard.go:		res := kcpClient.RESTClient().Get().AbsPath("/readyz").Do(ctx)
```